core/state/state_object.go

```diff
@@ -395,6 +395,63 @@ func (s *stateObject) CommitTrie(db Database) (*trie.NodeSet, error) {
 	return nodes, err
 }
 
+// DeleteTrie the storage trie of the object from db.
+func (s *stateObject) DeleteTrie(db Database) (*trie.NodeSet, error) {
+	// Track the amount of time wasted on iterating and deleting the storage trie
+	if metrics.EnabledExpensive {
+		defer func(start time.Time) { s.db.StorageDeletes += time.Since(start) }(time.Now())
+	}
+	// Open the original account trie from database
+	acctTrie, err := db.OpenTrie(s.db.originalRoot)
+	if err != nil {
+		return nil, err
+	}
+	// Load the original account without any mutation
+	origin, err := acctTrie.TryGetAccount(s.address.Bytes())
+	if err != nil {
+		return nil, err
+	}
+	// If the contract was not present in the state or the origin contract storage
+	// was empty, do nothing here. There are a few possible cases here:
+	// - the contract storage was empty and is destructed in this block
+	// - the contract is created and destructed in this block
+	if origin == nil || origin.Root == emptyRoot {
+		return nil, nil
+	}
+	// Open the original storage trie from database without any mutation. Because in
+	// the same block, the contract storage may be modified, and we only want to wipe
+	// the storage corresponding to the originalRoot in database.
+	stTrie, err := db.OpenStorageTrie(s.db.originalRoot, s.addrHash, origin.Root)
+	if err != nil {
+		return nil, err
+	}
+	// It can be an attack vector when iterating a huge contract. Stop collecting
+	// in case the accumulated nodes reach the threshold. It's fine to not clean
+	// up the dangling trie nodes since they are non-accessible dangling nodes
+	// anyway.
+	var (
+		paths [][]byte
+		blobs [][]byte
+		size  common.StorageSize
+		iter  = stTrie.NodeIterator(nil)
+	)
+	for iter.Next(true) {
+		if iter.Hash() == (common.Hash{}) {
+			continue
+		}
+		path, blob := common.CopyBytes(iter.Path()), common.CopyBytes(iter.NodeBlob())
+		paths = append(paths, path)
+		blobs = append(blobs, blob)
+
+		// Pretty arbitrary number, approximately 1GB as the threshold
+		size += common.StorageSize(len(path) + len(blob))
+		if size > 1073741824 {
+			return nil, nil
```

Comment:
Do we want to ignore all the work done? I.e. drop the accumulated 1GB on the floor, not even deleting that part?

Reply:
Yep. Personally I feel even worse if we do partial deletion of storage. I would prefer to keep all of them so that the entire storage trie is just dangling and unaccessible.